c++/src/kj/compat/http.c++

```diff
@@ -683,6 +683,12 @@ void HttpHeaders::add(kj::String&& name, kj::String&& value) {
 
 void HttpHeaders::addNoCheck(kj::StringPtr name, kj::StringPtr value) {
   KJ_IF_MAYBE(id, table->stringToId(name)) {
+    // Assert instead of allowing a mysterious non-debug only segfault.
+    KJ_ASSERT(
+        id->id < indexedHeaders.size(), "Attempted to use an unknown indexed header id, "
```

Comment:
If I understand correctly, the only case this fails is in fact when the header table is not fully built, and `indexedHeaders` is empty, right? Because otherwise `table` should never contain IDs that aren't in `indexedHeaders`.

So we could write this as:

```
KJ_ASSERT(!indexedHeaders.empty(), "tried to use HttpHeaderTable before it was fully built");
```

I think this error message would be a bit clearer to the reader.

Reply:
Actually that doesn't seem right, it appears `indexedHeaders` is built up as it goes and so I can't see how this assertion would ever be hit.

Reply:
Sorry I'm confusing myself.

What would actually have to happen is the `HttpHeaders` object is constructed from the table before the table is built, and then _after that_ the desired header is added to the header table builder, and then the resulting ID is used to set the header in the `HttpHeaders` object.

But that seems kind of hard to do by accident.

Reply:
So I think this is what we want here:

```
KJ_REQUIRE(id->id < indexedHeaders.size(), "HttpHeaders object was constructed from HttpHeaderTable that wasn't fully built yet");
```

Reply:
I think it would have to be something like this:
```
KJ_ASSERT(!indexedHeaders.size() != table->idCount(), "tried to use HttpHeaderTable before it was fully built");
```
You could have added some but not all of the ids already. This is actually the failure case I believe I am experiencing. (We had added something like two header ids before we make the HttpHeaders instance and ten header ids after.)

The other thing that might be worth mentioning is that this variant of the assertion will fire even if you never would have used the unknown indexed headers. So if you register the header id "Foo" after making your `HttpHeaders` instance but never use it to parse "Foo", I think it will still fail the assertion. If you're okay with that, then I'd be happy to make the change.

Reply:
If it's actually an error to call `HttpHeader`'s constructor with an unbuilt table, I suppose it might be preferable to assert that directly in the constructor, but I guess that as the system is currently implemented, that would require adding a bit of state to tables to track whether they had been built or not.

I also notice that `HttpHeaders::set()`, `get()`, and `unset()` do a similarly unchecked index into `indexedHeaders` that might warrant a similar assert.

Reply:
> If it's actually an error to call HttpHeader's constructor with an unbuilt table, I suppose it might be preferable to assert that directly in the constructor, but I guess that as the system is currently implemented, that would require adding a bit of state to tables to track whether they had been built or not.

I think I'd call it a contract violation but I do agree that we can add an assertion with a small amount of state. It does potentially break users that otherwise would not have experienced a problem. Maybe a debug only assertion?

> I also notice that HttpHeaders::set(), get(), and unset() do a similarly unchecked index into indexedHeaders that might warrant a similar assert.

Sure, if you think it's worth it, I'm happy to add assertions there as well.

Reply:
> t does potentially break users that otherwise would not have experienced a problem.

Ehh I think the users potentially broken are already in a precarious position here, if they're constructing headers with unbuilt header tables. So I'd be happy with a hard error in the HttpHeaders constructor in this case.